services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java

```diff
@@ -31,12 +48,105 @@
-        throw new UnsupportedOperationException("SD-JWT Signing is not yet supported.");
+
+        DisclosureSpec.Builder disclosureSpecBuilder = DisclosureSpec.builder();
+        CredentialSubject credentialSubject = verifiableCredential.getCredentialSubject();
+        JsonNode claimSet = objectMapper.valueToTree(credentialSubject);
+        // put all claims into the disclosure spec, except the one to be kept visible
+        credentialSubject.getClaims()
+                .entrySet()
+                .stream()
+                .filter(entry -> !visibleClaims.contains(entry.getKey()))
+                .forEach(entry -> {
+                    if (entry instanceof List<?> listValue) {
+                        IntStream.range(0, listValue.size())
+                                .forEach(i -> disclosureSpecBuilder.withUndisclosedArrayElt(entry.getKey(), i, SdJwtUtils.randomSalt()));
+                    } else {
+                        disclosureSpecBuilder.withUndisclosedClaim(entry.getKey(), SdJwtUtils.randomSalt());
+                    }
+                });
+
+        // add the configured number of decoys
+        if (decoys != 0) {
+            IntStream.range(0, decoys)
+                    .forEach(i -> disclosureSpecBuilder.withDecoyClaim(SdJwtUtils.randomSalt()));
+        }
+
+        ObjectNode rootNode = claimSet.withObject("");
+        rootNode.put(ISSUER_CLAIM, issuerDid);
+
+        // Get the issuance date from the credential. Since nbf is mandatory, we set it to the current time if not
+        // provided
+        long iat = Optional.ofNullable(verifiableCredential.getIssuanceDate())
+                .map(issuanceDate -> issuanceDate.toInstant().getEpochSecond())
+                .orElse((long) timeProvider.currentTimeSeconds());
+        rootNode.put(NOT_BEFORE_CLAIM, iat);
+        if (verifiableCredential.getType() == null || verifiableCredential.getType().size() != 1) {
+            throw new SigningServiceException("SD-JWT only supports single type credentials.");
+        }
+        rootNode.put(VERIFIABLE_CREDENTIAL_TYPE_CLAIM, verifiableCredential.getType().get(0));
+        rootNode.put(SELECTIVE_DISCLOSURE_ALGORITHM_CLAIM, hashAlgorithm);
+        rootNode.put(CREDENTIAL_ID_CLAIM, JwtSigningService.createCredentialId(verifiableCredential));
+
+        SdJwt sdJwt = SdJwt.builder()
+                .withDisclosureSpec(disclosureSpecBuilder.build())
+                .withClaimSet(claimSet)
+                .withSigner(signatureSignerContext)
+                .withHashAlgorithm(hashAlgorithm)
+                .withJwsType(tokenType)
+                .build();
+
+        StringJoiner tokenJoiner = new StringJoiner("~");
+        // start with the actual credential
+        tokenJoiner.add(sdJwt.getIssuerSignedJWT().getJwsString());
+        // append the disclosures
+        sdJwt.getDisclosures().forEach(tokenJoiner::add);
+        return tokenJoiner.toString();
```

Comment:
Why not use this?
```suggestion
        return sdJwt.toSdJwtString();
```

Reply:
Because I somehow overread that method;) Thanks for mentioning